src/rust/engine/options/src/args.rs

```diff
@@ -3,95 +3,140 @@
 
 use std::env;
 
-use super::id::{NameTransform, OptionId, Scope};
+use super::id::{is_valid_scope_name, NameTransform, OptionId, Scope};
 use super::{DictEdit, OptionsSource};
-use crate::parse::{expand, expand_to_dict, expand_to_list, Parseable};
+use crate::parse::{expand, expand_to_dict, expand_to_list, ParseError, Parseable};
 use crate::ListEdit;
-use std::collections::HashMap;
-
-pub struct Args {
-    pub(crate) args: Vec<String>,
+use core::iter::once;
+use itertools::{chain, Itertools};
+
+#[derive(Debug)]
+struct Arg {
+    context: Scope,
+    flag: String,
+    value: Option<String>,
 }
 
-#[derive(PartialEq)]
-enum Negate {
-    True,
-    False,
-}
+impl Arg {
+    /// Checks if this arg's flag is equal to the provided strings concatenated with dashes.
+    /// E.g., "--foo-bar" matches ["-", "foo", "bar"].
+    fn flag_match<'a>(&self, dash_separated_strs: impl Iterator<Item = &'a str>) -> bool {
+        #[allow(unstable_name_collisions)]
+        // intersperse is provided by itertools::Itertools, but is also in the Rust nightly
+        // as an experimental feature of standard Iterator. If/when that becomes standard we
+        // can use it, but for now we must squelch the name collision.
+        itertools::equal(
+            self.flag.chars(),
+            dash_separated_strs
+                .map(str::chars)
+                .intersperse("-".chars())
+                .flatten(),
+        )
+    }
 
-impl Args {
-    pub fn new(args: Vec<String>) -> Self {
-        Self { args }
+    /// Checks if this arg provides a value for the specified option, either negated or not.
+    fn _matches(&self, id: &OptionId, negate: bool) -> bool {
+        fn prefix<'a>(negate: bool) -> impl Iterator<Item = &'a str> {
+            if negate {
+                once("--no")
+            } else {
+                once("-")
+            }
+        }
+        self.flag_match(chain![
+            prefix(negate),
+            once(id.scope.name()),
+            id.name_components_strs()
+        ]) || (self.context == id.scope
+            && self.flag_match(chain![prefix(negate), id.name_components_strs()]))
+            || (if let Some(sn) = &id.short_name {
+                self.flag_match(chain![once(""), once(sn.as_ref())])
+            } else {
+                false
+            })
```

Comment:
We can't deconstruct because we don't know where the scope ends and the option name begins. E.g., `--foo-bar-baz`: is that option `baz` in scope `foo-bar` or option `bar-baz` in scope `foo`? Plus, if there's a `no-` in there things get even more complicated.  I'll add some comments to clarify the logic here.